ddht/v5_1/alexandria/advertisements.py

```diff
@@ -0,0 +1,195 @@
+            expires_at = datetime.datetime.utcnow().replace(microsecond=0) + ONE_DAY
+        content_id = content_key_to_content_id(content_key)
+        signature = create_advertisement_signature(
+            content_id, hash_tree_root, expires_at, private_key,
+        )
+        return cls(
+            content_key=content_key,
+            hash_tree_root=hash_tree_root,
+            expires_at=expires_at,
+            signature_v=signature.v,
+            signature_r=signature.r,
+            signature_s=signature.s,
+        )
+
+    def verify(self, public_key: keys.PublicKey) -> None:
+        verify_advertisement_signature(
+            content_id=self.content_id,
+            hash_tree_root=self.hash_tree_root,
+            expires_at=self.expires_at,
+            signature=self.signature,
+            public_key=public_key,
+        )
+
+    def to_sedes_payload(self) -> Tuple[ContentKey, Hash32, int, int, int, int]:
+        return (
+            self.content_key,
+            self.hash_tree_root,
+            int(self.expires_at.timestamp()),
+            self.signature_v,
+            self.signature_r,
+            self.signature_s,
+        )
+
+    @classmethod
+    def from_sedes_payload(
+        cls, payload: Tuple[ContentKey, Hash32, int, int, int, int],
+    ) -> "Advertisement":
+        (
+            content_key,
+            hash_tree_root,
+            expires_at_timestamp,
+            signature_v,
+            signature_r,
+            signature_s,
+        ) = payload
+        expires_at = datetime.datetime.fromtimestamp(expires_at_timestamp)
+        return cls(
+            content_key,
+            hash_tree_root,
+            expires_at,
+            signature_v,
+            signature_r,
+            signature_s,
+        )
+
+
+ADVERTISEMENT_FIXED_SIZE = sum(
+    (
+        4,  # length prefix for the whole advertisement
+        4,  # length prefix for `content_key`
+        65,  # signature
+        5,  # expiration
+        32,  # hash_tree_root
+    )
+)
+
+
+def _get_partition_indices(
+    encoded_sizes: Sequence[int], max_payload_size: int
+) -> Iterable[int]:
+    cumulative_sizes = accumulate(operator.add, encoded_sizes)
+    offset = 0
+    yield 0
+    for idx, (last_size, size) in enumerate(
+        sliding_window(2, cons(0, cumulative_sizes))
+    ):
+        if size - offset > max_payload_size:
+            offset = last_size
+            yield idx
+    yield len(encoded_sizes)
+
+
+@to_tuple
+def partition_advertisements(
+    advertisements: Sequence[Advertisement], max_payload_size: int,
+) -> Iterable[Tuple[Advertisement, ...]]:
+    encoded_sizes = tuple(
+        len(advertisement.content_key) + ADVERTISEMENT_FIXED_SIZE
+        for advertisement in advertisements
+    )
+    partition_indices = _get_partition_indices(encoded_sizes, max_payload_size)
+    for left, right in sliding_window(2, partition_indices):
+        yield tuple(advertisements[left:right])
```

Comment:
If I remember correctly you wrote very similar logic in a previous PR. It might simplify things to pull the "break this sequence of items into chunks of at most size n" into a helper method both can call

Reply:
Conceptually yes, we do the same thing with ENR records

https://github.com/ethereum/ddht/blob/fd5dd92933d336b49383808c2c78eb7e28159b0c/ddht/enr.py#L10-L46

But the logic is different enough that they don't DRY well.